### DISCUSSION TITLE
Loading translations on transitions

### DISCUSSION THREAD

**Discussion by BenoitAverty** [sergiodxa/remix-i18next#78] [Q&A] [answered]
Hey,

I've paused the dev of my remix app for a few months and I've just upgraded remix-i18next to v4. I've seen on the readme that you switched to a `handle` export in the routes to specify the namespace, and that the translations are no longer loaded from the loader. As we've discussed before, I think this is a better DX.

However, this means the translations from the client are loaded by the http backend, which is fine, but as far as I know there is still no way to tell remix to wait for the translations before navigating to the new route. This means that if the translations take long to download by the i18next-htt-backend then there is a time when we see the keys instead of the translated text.

Is there a way to avoid that in the new version ?

Thanks in advance.

**Comment by sergiodxa** [comment, marked as answer]
The only way is to preload them before clicking a link, or preload them with a LinksFunction in the route but you'll need to preload all translations due the lack of loader data to know the locale.

To avoid a flash of translation keys you will have to check if useTranslation is ready by doing

```ts
let { t, ready } = useTranslation("ns")
if (!ready) return <Loading />
return <p>{t("key")}</p>
```

I think for most apps, using a single namespace for all translations is gonna be simpler, or loading all translation files up front, maybe importing them, at least until Remix exposes a way a library can ask Remix to wait for some asset to load before doing a client-side navigation.

For larger apps when this is not possible the only solution is to check for the `ready` boolean for now.
